File: macOS/RCForb/Sources/Protocol/ProtocolConstants.swift
import Foundation

enum ControlByte {
    static let HEARTBEAT: UInt8 = 0x00
    static let PTT: UInt8 = 0x03
    static let PINGPONG: UInt8 = 0x05
    static let PTT_OFF: UInt8 = 0x06
    static let KEY_OFF: UInt8 = 0xFB
    static let KEY_ON: UInt8 = 0xFC
    static let USERIN: UInt8 = 0xFD
    static let USEROUT: UInt8 = 0xFE
    static let UTF8STRING: UInt8 = 0xFF
}

let HEARTBEAT_TIMEOUT_MS: Int = 15_000
let HEARTBEAT_INTERVAL_MS: Int = 4_000
let PING_INTERVAL_MS: Int = 1_000
let DEFAULT_CMD_PORT: Int = 4525
let DEFAULT_AUDIO_PORT: Int = 4524

let OPUS_SAMPLE_RATE: Int = 48000
let OPUS_CHANNELS: Int = 1
let OPUS_BITRATE: Int = 24000
let OPUS_FRAME_MS: Int = 20
let OPUS_FRAME_SIZE: Int = (OPUS_SAMPLE_RATE * OPUS_FRAME_MS) / 1000 // 960

enum PacketType {
    case audio
    case command
    case heartbeat
    case pingpong
    case pttOn
    case pttOff
    case keyOn
    case keyOff
    case userIn
    case userOut
}

func classifyPacket(_ data: Data) -> PacketType {
    if data.isEmpty { return .heartbeat }
    let firstByte = data[data.startIndex]

    if firstByte == ControlByte.UTF8STRING && data.count > 1 { return .command }
    if data.count == 1 {
        switch firstByte {
        case ControlByte.HEARTBEAT: return .heartbeat
        case ControlByte.PINGPONG: return .pingpong
        case ControlByte.PTT: return .pttOn
        case ControlByte.PTT_OFF: return .pttOff
        case ControlByte.KEY_ON: return .keyOn
        case ControlByte.KEY_OFF: return .keyOff
        case ControlByte.USERIN: return .userIn
        case ControlByte.USEROUT: return .userOut
        default: break
        }
    }
    return .audio
}

File: android/app/src/main/java/com/rcforb/android/protocol/ProtocolConstants.kt
package com.rcforb.android.protocol

object ControlByte {
    const val HEARTBEAT: Byte = 0x00
    const val PTT: Byte = 0x03
    const val PINGPONG: Byte = 0x05
    const val PTT_OFF: Byte = 0x06
    const val KEY_OFF: Byte = 0xFB.toByte()
    const val KEY_ON: Byte = 0xFC.toByte()
    const val USERIN: Byte = 0xFD.toByte()
    const val USEROUT: Byte = 0xFE.toByte()
    const val UTF8STRING: Byte = 0xFF.toByte()
}

const val HEARTBEAT_TIMEOUT_MS = 15_000L
const val HEARTBEAT_INTERVAL_MS = 4_000L
const val PING_INTERVAL_MS = 1_000L
const val DEFAULT_CMD_PORT = 4525
const val DEFAULT_AUDIO_PORT = 4524

const val OPUS_SAMPLE_RATE = 48000
const val OPUS_CHANNELS = 1
const val OPUS_BITRATE = 24000
const val OPUS_FRAME_MS = 20
const val OPUS_FRAME_SIZE = (OPUS_SAMPLE_RATE * OPUS_FRAME_MS) / 1000 // 960

enum class PacketType {
    AUDIO, COMMAND, HEARTBEAT, PINGPONG,
    PTT_ON, PTT_OFF, KEY_ON, KEY_OFF,
    USER_IN, USER_OUT
}

fun classifyPacket(data: ByteArray): PacketType {
    if (data.isEmpty()) return PacketType.HEARTBEAT
    val firstByte = data[0]

    if (firstByte == ControlByte.UTF8STRING && data.size > 1) return PacketType.COMMAND
    if (data.size == 1) {
        return when (firstByte) {
            ControlByte.HEARTBEAT -> PacketType.HEARTBEAT
            ControlByte.PINGPONG -> PacketType.PINGPONG
            ControlByte.PTT -> PacketType.PTT_ON
            ControlByte.PTT_OFF -> PacketType.PTT_OFF
            ControlByte.KEY_ON -> PacketType.KEY_ON
            ControlByte.KEY_OFF -> PacketType.KEY_OFF
            ControlByte.USERIN -> PacketType.USER_IN
            ControlByte.USEROUT -> PacketType.USER_OUT
            else -> PacketType.AUDIO
        }
    }
    return PacketType.AUDIO
}
